### PR TITLE
virtual_keyboard: Release keys on destroy

### LINKED ISSUE #2034: virtual-keyboard: release unreleased keys
Follow https://github.com/swaywm/sway/pull/4932#issuecomment-586108940.

It's not restricted to IME. Any application aborted may leave some unreleased virtual_keyboard key events.  I think there should be similar thing to https://github.com/swaywm/wlroots/blob/master/types/wlr_keyboard_group.c#L203-L219.

As it is using wlr_keyboard, so pressed keys are recorded. So keyboard_group can handle unreleased keys, then why can not virtual_keyboard?

### PR DESCRIPTION
**Pull request by tadeokondrak**
Fixes #2034.

### CODE REVIEW

**Comment by emersion** [comment]
Hm, but maybe this applies to physical keyboards as well? What happens if you hold a key and disconnect your keyboard? Same applies to the keyboards for the Wayland/X11 backends.

Maybe this logic should be part of [`wlr_keyboard_destroy`](https://github.com/swaywm/wlroots/blob/7b50f5d3249173d93736d6e2a3c6f61e86b367bd/types/wlr_keyboard.c#L127)?

**Comment by tadeokondrak** [comment]
> Hm, but maybe this applies to physical keyboards as well? What happens if you hold a key and disconnect your keyboard?

It doesn't keep repeating, I'm not sure what handles this.

> Wayland backend

Is the parent Wayland compositor leaving a key pressed when removing the keyboard capability from a seat something that clients should handle?

> X11 backend

It only has one seat, so keyboards can't be removed during runtime.

File: include/wlr/interfaces/wlr_keyboard.h [hunk 1]
```diff
@@ -19,6 +19,7 @@ struct wlr_keyboard_impl {
 
 void wlr_keyboard_init(struct wlr_keyboard *keyboard,
 		const struct wlr_keyboard_impl *impl);
+void wlr_keyboard_release_keys(struct wlr_keyboard *keyboard);
```
Comment (tadeokondrak) [inline include/wlr/interfaces/wlr_keyboard.h:22]: Maybe this shouldn't be a public function. Not sure what to do instead, though.

File: types/wlr_keyboard.c [hunk 2]
```diff
@@ -124,6 +126,22 @@ void wlr_keyboard_init(struct wlr_keyboard *kb,
...
+			.keycode = kb->keycodes[0],
+			.update_state = false,
+			.state = WL_KEYBOARD_KEY_STATE_RELEASED,
+		};
+		wlr_keyboard_notify_key(kb, &event);
+	}
+}
+
 void wlr_keyboard_destroy(struct wlr_keyboard *kb) {
 	if (kb == NULL) {
```
Comment (jaens) [inline types/wlr_keyboard.c:146]: Could `wlr_keyboard_release_keys` be called inside `wlr_keyboard_destroy` instead?
That way it could be private.

Comment (emersion) [reply types/wlr_keyboard.c:146]: That would be best indeed. I'm not a fan of having keyboard-specific stuff inside the common code.

Comment (tadeokondrak) [reply types/wlr_keyboard.c:146]: `wlr_keyboard_destroy` is called after the `destroy` signal on the `wlr_input_device` is emitted.
If the keys are released after that signal, the original issue isn't fixed anymore.

**Comment by emersion** [comment]
I wonder if instead we should keep the status quo and make it so destroying a keyboard implicitly releases all keys.

**Comment by emersion** [comment]
Superseded by https://github.com/swaywm/wlroots/pull/2918

**Pull request closed without merging**